go/processor/floria/calls.go

```diff
@@ -16,62 +16,70 @@ import (
 
-	blockParameters := tosca.BlockParameters{}
-
-	transactionParameters := tosca.TransactionParameters{
-		Origin:     transaction.Sender,
-		GasPrice:   transaction.GasPrice,
-		BlobHashes: []tosca.Hash{},
+	if r.depth > CallCreateDepth {
+		return tosca.CallResult{}, nil
 	}
+	r.depth++
+	defer func() { r.depth-- }()
 
-	codeHash := context.GetCodeHash(*transaction.Recipient)
-	code := context.GetCode(*transaction.Recipient)
+	codeHash := r.GetCodeHash(parameters.Recipient)
+	code := r.GetCode(parameters.Recipient)
 
 	params := tosca.Parameters{
-		BlockParameters:       blockParameters,
-		TransactionParameters: transactionParameters,
-		//Context:               runContext, todo implement
-		Kind:      tosca.Call,
-		Static:    false,
-		Depth:     0, // todo add depth check
-		Gas:       gas,
-		Recipient: *transaction.Recipient,
-		Sender:    transaction.Sender,
-		Input:     transaction.Input,
-		Value:     transaction.Value,
-		CodeHash:  &codeHash,
-		Code:      code,
+		BlockParameters:       r.blockParams,
+		TransactionParameters: r.transactionParameters,
+		Context:               r,
+		Kind:                  tosca.Call,
+		Static:                false,
+		Depth:                 r.depth - 1, // depth is already incremented
+		Gas:                   parameters.Gas,
+		Recipient:             parameters.Recipient,
+		Sender:                parameters.Sender,
+		Input:                 parameters.Input,
+		Value:                 parameters.Value,
+		CodeHash:              &codeHash,
+		Code:                  code,
 	}
 
-	snapshot := context.CreateSnapshot()
-	if err := transferValue(transaction, context); err != nil {
-		context.RestoreSnapshot(snapshot)
-		return tosca.Result{}, nil
+	snapshot := r.CreateSnapshot()
+	if err := transferValue(r, parameters.Value, parameters.Sender, parameters.Recipient); err != nil {
+		r.RestoreSnapshot(snapshot)
+		return tosca.CallResult{}, nil
 	}
 
-	result, err := interpreter.Run(params)
+	result, err := r.interpreter.Run(params)
 	if err != nil || !result.Success {
-		context.RestoreSnapshot(snapshot)
+		r.RestoreSnapshot(snapshot)
 	}
 
-	return result, err
+	return tosca.CallResult{
+		Output:    result.Output,
+		GasLeft:   result.GasLeft,
+		GasRefund: result.GasRefund,
+		Success:   result.Success,
+	}, err
 }
 
-func transferValue(transaction tosca.Transaction, context tosca.TransactionContext) error {
-	senderBalance := context.GetBalance(transaction.Sender)
-	transferValue := transaction.Value
+func transferValue(
+	context tosca.TransactionContext,
+	value tosca.Value,
+	sender tosca.Address,
+	recipient tosca.Address,
+) error {
+	senderBalance := context.GetBalance(sender)
+	transferValue := value
 	if senderBalance.Cmp(transferValue) < 0 {
 		return fmt.Errorf("insufficient balance: %v < %v", senderBalance, transferValue)
 	}
 
 	senderBalance = tosca.Sub(senderBalance, transferValue)
-	context.SetBalance(transaction.Sender, senderBalance)
+	context.SetBalance(sender, senderBalance)
```

Comment:
is negative balance allowed?

Reply:
No, there is a check whether the sender has sufficient balance.